src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp

```diff
@@ -139,15 +355,88 @@ bool ControllerScriptEngineLegacy::initialize() {
+                watchFilePath(it.next());
+            }
+            watchFilePath(path);
+            auto pQmlEngine = std::dynamic_pointer_cast<QQmlEngine>(m_pJSEngine);
+            pQmlEngine->addImportPath(path);
+            qCWarning(m_logger) << pQmlEngine->importPathList();
+        }
+    } else if (!m_modules.isEmpty()) {
+        qCWarning(m_logger) << "Controller mapping has QML library definitions but no "
+                               "QML files to use it. Ignoring.";
+    }
+
+    // If we encounter a failure while loading a scene, we will need to properly
+    // stop the screen threads before shutting down.
+    bool sceneBindingHasFailure = false;
+#endif
     for (const LegacyControllerMapping::ScriptFileInfo& script : std::as_const(m_scriptFiles)) {
-        if (!evaluateScriptFile(script.file)) {
-            shutdown();
-            return false;
+#ifdef MIXXX_USE_QML
+        if (script.type == LegacyControllerMapping::ScriptFileInfo::Type::JAVASCRIPT) {
+#endif
+            if (!evaluateScriptFile(script.file)) {
+                shutdown();
+                return false;
+            }
+            if (!script.identifier.isEmpty()) {
+                m_scriptFunctionPrefixes.append(script.identifier);
+            }
+#ifdef MIXXX_USE_QML
+        } else {
+            if (script.identifier.isEmpty()) {
+                while (!availableScreens.isEmpty()) {
+                    QString screenIdentifier(availableScreens.firstKey());
+                    if (!bindSceneToScreen(script,
+                                screenIdentifier,
+                                availableScreens.take(screenIdentifier))) {
+                        sceneBindingHasFailure = true;
+                    }
+                }
+            } else {
+                if (!availableScreens.contains(script.identifier)) {
+                    qCCritical(m_logger) << "Not screen" << script.identifier << "found!";
+
+                    sceneBindingHasFailure = true;
+                    break;
+                }
+                if (!bindSceneToScreen(script,
+                            script.identifier,
+                            availableScreens.take(script.identifier))) {
+                    sceneBindingHasFailure = true;
+                }
+            }
         }
-        if (!script.functionPrefix.isEmpty()) {
-            m_scriptFunctionPrefixes.append(script.functionPrefix);
+    }
+
+    if (!availableScreens.isEmpty()) {
+        if (!sceneBindingHasFailure) {
+            qCWarning(m_logger)
+                    << "Found screen with no QML scene able to run on it. Ignoring"
+                    << availableScreens.size() << "screens";
+        }
+
+        while (!availableScreens.isEmpty()) {
+            std::shared_ptr<ControllerRenderingEngine> orphanScreen(
+                    nullptr, &delete_later<ControllerRenderingEngine>);
+            availableScreens.take(availableScreens.firstKey()).swap(orphanScreen);
```

Comment:
@Swiftb0y this is the `swap` that replaced the `std::move`

Reply:
just so I understand, what is this block accomplishing? it looks like it is moving items from the availablescreens list and making sure they get deleted because they are orphans?  why is this needed instead of letting availablescreens go out of scope?

Reply:
Because `orphanScreen` has now couple of component with thread affinity to its own thread (OpenGL, attached the screen thread, OffscreenSurface, needed to be deleted by GUI thread, ...). By calling `deleteLater`, which make sure that its own event loop will take care of the correct destruction of these component. An alternative could be the join/wait the thread here.

Reply:
Okay, thinking twice, this is ugly hack and shouldn't be there. I'm using the `stop` method now, that does a proper `wait`. Tested and working just fine.